File: 212. Word Search II/212. Word Search II.java
import java.util.*;
class Solution {
    private class TrieNode{
        Map<Character, TrieNode> children = new HashMap<>();
        String word;
    }
    private TrieNode root = new TrieNode();
    private List<String> res = new ArrayList<>();
    
    public List<String> findWords(char[][] board, String[] words) {
        for(String word : words){
            buildTrie(word);
        }
        for(int i = 0; i < board.length; i++){
            for(int j = 0; j < board[0].length; j++){
                backtracking(board, i , j, root);
            }
        }
        return res;
    }
    
    private void backtracking(char[][] board, int i, int j, TrieNode p){
        char c = board[i][j];
        p = p.children.get(c);
        if(c == '#' || p == null) return;
        if(p.word != null) {
            res.add(p.word);
            p.word = null;
            if(p.children.isEmpty()) return;
        };
        
        board[i][j] = '#';
        if(i > 0) backtracking(board, i - 1, j ,p);
        if(j > 0) backtracking(board, i, j -1, p);
        if(i < board.length - 1) backtracking(board, i + 1, j, p);
        if(j < board[0].length - 1) backtracking(board, i, j + 1, p);
        board[i][j] = c;
        
    }
    
    private void buildTrie(String str){
        TrieNode cur = root;
        for(char ch : str.toCharArray()){
            if(!cur.children.containsKey(ch)){
                cur.children.put(ch, new TrieNode());
            }
            cur = cur.children.get(ch);
        }
        cur.word = str;
        
    }
}

File: 212. Word Search II/212. Word Search II.js
/**
 * @param {character[][]} board
 * @param {string[]} words
 * @return {string[]}
 */
class TrieNode {
  constructor() {
    this.children = new Map();
    this.word = null;
  }
}
var findWords = function(board, words) {
  const root = buildTrie(words);

  const res = [];
  for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[0].length; j++) {
      dfs(board, i, j, root, res);
    }
  }
  return res;
};

var dfs = function(board, i, j, p, res) {
  let c = board[i][j];
  if (!p.children.has(c) || c == "#") return;

  p = p.children.get(c);
  if (p.word != null) {
    res.push(p.word);
    p.word = null;
    if (p.children.size == 0) return;
  }

  board[i][j] = "#";
  if (i > 0) dfs(board, i - 1, j, p, res);
  if (j > 0) dfs(board, i, j - 1, p, res);
  if (i < board.length - 1) dfs(board, i + 1, j, p, res);
  if (j < board[0].length - 1) dfs(board, i, j + 1, p, res);

  board[i][j] = c;
};

var buildTrie = function(words) {
  const root = new TrieNode();
  words.forEach(str => {
    let cur = root;
    for (let c of str) {
      if (!cur.children.has(c)) {
        cur.children.set(c, new TrieNode());
      }
      cur = cur.children.get(c);
    }
    cur.word = str;
  });
  return root;
};
